internal/utils/time.go

```diff
@@ -59,3 +59,53 @@ func RunOnInterval(ctx context.Context, fn func(), interval time.Duration) {
 		}
 	}()
 }
+
+// RepeatPolicy repeat behavior for RunOnIntervals Function
+type RepeatPolicy int
+
+const (
+	// RepeatAll repeats all intervals
+	RepeatAll RepeatPolicy = iota
+	// RepeatLast repeats only the last interval
+	RepeatLast
+	// RepeatNone does not repeat
+	RepeatNone
+)
+
+// RunOnIntervals the given function once on the specified intervals, and
+// repeat according to the supplied RepeatPolicy.
+func RunOnIntervals(ctx context.Context, fn func(), intervals []time.Duration, repeatPolicy RepeatPolicy) {
+	if len(intervals) < 1 {
+		return
+	}
+	// copy intervals
+	intv := intervals[:]
+
+	// initialize timer
+	timer := time.NewTimer(intv[0])
+	go func() {
+		defer timer.Stop()
+
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-timer.C:
+
+				if len(intv) > 1 {
+					// advance the interval
+					intv = intv[1:]
+				} else {
```

Comment:
Doesn't this skip the last interval for repeat policy of `None` or `All` since you are resetting the time _after_ the slice has been pared down?  So when there is one interval remaining in the slice, the `else` block will be hit and then it will return or reset `intv` with the last element ignored.

Maybe add some unit tests to test all of the various policies.

Reply:
No, it does execute.  The timer is initialized with intv[0].  We do not pare down the interval array until after the timer fires.  Then we evaluate based on the current length of the interval what to do.  If the len(intv) > 1 we advance and reset the timer.  If the len(intv) == 1 it means we've already run the last interval through the timer.  That's when the reset policies are evaluated.  I could change the first condition to `if (len(intv) >= 2){} else{}` if that's clearer?

Reply:
I think I understand what you're saying here, yes you're right because the function is executed after the evaluation of the repeat policy it would effectively skip it.